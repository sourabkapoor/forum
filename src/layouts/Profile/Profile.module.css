.profile {
  width: 100%;
}
.profile__button {
  border: 1px solid #cecece;
  background-color: white;
  border-radius: 4px;
  color: gray;
  padding: 20x 0;
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 4px;
  margin-left: 8px;
}

.profile__content{
    text-align: left;
  width: 110px;
}

.profile__username {
    font-size: 14px;
    font-weight: 500;
    line-height: 16px;
    white-space: nowrap;
    color: var(--secondary-color);
}

.profile__karma {
    font-size: 12px;
    font-weight: 500;
    line-height: 16px;
    color: #8a8a8a;
}

.profile__button:hover {
  border: 1px solid #343536;
  cursor: pointer;
  border-radius: 4px;
}

.profile__details {
    display: flex;
    align-items: center;
}

.profile__avatar {
    width: 24px;
    height: 24px;
    margin-right: 5px;
}

.profile__button-icon {
  font-size: 20px;
  font-weight: 400;
  height: 20px;
  line-height: 20px;
  vertical-align: middle;
  width: 20px;
}

@media(max-width: 1210px) {
  .profile__content {
    display: none;
  }
}

@media(max-width: 640px) {
  .profile__button {
    margin: 0;
    margin-left: 2px;
  }
}
