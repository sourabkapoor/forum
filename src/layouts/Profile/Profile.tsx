import React, { useEffect, useState } from "react";
import { FaRegUser } from "react-icons/fa";
import { IoIosArrowDown } from "react-icons/io";
import styles from "./Profile.module.css";
import onlineStatus from "../../assets/online-status.svg";
import CSSModules from "react-css-modules";
import Dropdown from "../Dropdown/Dropdown";
import AccountCircleIcon from '@mui/icons-material/AccountCircle';

import {
  collection,
  doc,
  DocumentData,
  getDoc,
  getDocs,
  query,
  where,
} from "firebase/firestore";
import { db, getUser, getUserId, getUserName } from "../../firebase";

interface Props {
  isLoggedIn: boolean;
}

const Profile: React.FC<Props> = ({ isLoggedIn }) => {
  const [userPosts, setuserPosts] = useState(0)

  useEffect(() => {
    async function getuserPosts() {
      const userPostsRef = collection(db, "posts");
      const q = query(userPostsRef, where("userId", "==", getUserId()));
      const userPostsDoc = await getDocs(q);
      setuserPosts(userPostsDoc.docs.length)
    }
    if(isLoggedIn)
      getuserPosts()
  }, [isLoggedIn])

  return (
    <div styleName="profile">
      {isLoggedIn ? (
          <button styleName="profile__button">
          <span styleName="profile__details">
            <AccountCircleIcon styleName="profile__avatar" />
            <div styleName="profile__content">
              <p styleName="profile__username">{getUserName()}</p>
              {userPosts !== 0 && <p styleName="profile__karma">{userPosts + " Posts"}</p>}
            </div>
          </span>
          
          <IoIosArrowDown styleName="profile__button-icon" />
        </button>
          ) : (
            null
          )}
    </div>
  );
};

export default CSSModules(Profile, styles, {
  allowMultiple: true,
  handleNotFoundStyleName: "ignore",
});
