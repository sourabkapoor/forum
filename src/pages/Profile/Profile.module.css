.main {
  display: flex;
  justify-content: center;
  padding: 20px 24px;
  column-gap: 24px;
  width: 100%;
}
.profile-container {
  animation: fadeIn ease 2s;
  width: 100%;
}
@keyframes fadeIn {
  from {
    opacity: 0%;
  }
  to {
    opacity: 100%;
  }
}

@media (max-width: 1070px) {
  .main {
    flex-direction: column-reverse;
  }
  .content {
    width: 100%;
  }
  .main {
    padding: 0;
    padding-top: 24px;
  }
}
