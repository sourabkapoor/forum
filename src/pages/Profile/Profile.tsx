/* eslint-disable eqeqeq */
import {
  collection,
  doc,
  DocumentData,
  getDoc,
  getDocs,
  query,
  where,
} from "firebase/firestore";
import { useEffect, useReducer, useRef, useState } from "react";
import CSSModules from "react-css-modules";
import { db, getUser, getUserId, getUserName } from "../../firebase";
import Filter from "../../components/Filter/Filter";
import Comments from "../Subreddit/Comments/Comments";
import Posts from "../Subreddit/Posts/Posts";
import Header from "./Header/Header";
import styles from "./Profile.module.css";
import UserCard from "./UserCard/UserCard";
import { useAppSelector, useFilter } from "../../hooks/hooks";
import Card from "../../components/Card/Card";
import ProfileNotFound from "./ProfileNotFound/ProfileNotFound";
import { Navigate, useNavigate } from "react-router-dom";
import { displayPartsToString } from "typescript";
import {
  selectAuthStatus,
  toggleSignInModal,
} from "../../features/auth/authSlice";
import Main from "../../layouts/Main/Main";
import Aside from "../../layouts/Aside/Aside";
import { createHash } from "crypto";

interface State {
  posts: boolean;
  comments: boolean;
  upvotes: boolean;
  downvotes: boolean;
  saved: boolean;
}

interface Action {
  type: string;
}

function reducer(state: State, action: Action) {
  switch (action.type) {
    case "POSTS":
      return {
        posts: true,
        comments: false,
        upvotes: false,
        downvotes: false,
        saved: false,
      };
    case "COMMENTS":
      return {
        posts: false,
        comments: true,
        upvotes: false,
        downvotes: false,
        saved: false,
      };
    case "UPVOTES":
      return {
        posts: false,
        comments: false,
        upvotes: true,
        downvotes: false,
        saved: false,
      };
    case "DOWNVOTES":
      return {
        posts: false,
        comments: false,
        upvotes: false,
        downvotes: true,
        saved: false,
      };
    case "SAVED":
      return {
        posts: false,
        comments: false,
        upvotes: false,
        downvotes: false,
        saved: true,
      };
    default:
      return state;
  }
}

const Profile: React.FC = () => {
  const filter = useFilter();

  const [userPosts, setUserPosts] = useState<DocumentData | undefined>();

  const [userComments, setUserComments] = useState<DocumentData | undefined>();

  const [commentsPostId, setCommentsPostId] = useState<string | undefined>();

  const [upVotesPosts, setUpVotesPosts] = useState<DocumentData | undefined>();

  const [savedPosts, setSavedPosts] = useState<DocumentData | undefined>();

  function addPosts() {}

  const [downVotesPosts, setDownVotesPosts] = useState<
    DocumentData | undefined
  >();

  const isLoggedIn = useAppSelector(selectAuthStatus);
  const navigate = useNavigate();

  const [activeSection, dispatch] = useReducer(reducer, {
    posts: true,
    comments: false,
    upvotes: false,
    downvotes: false,
    saved: false,
  });

  useEffect(() => {
    async function getUserPosts() {
      if (!isLoggedIn) {
        dispatch(toggleSignInModal());
        return;
      }
      try {
        const userPostsRef = collection(db, "posts");

        const q = query(userPostsRef, where("userId", "==", getUserId()));

        const userPostsDoc = await getDocs(q);
        setUserPosts(userPostsDoc.docs);
      } catch (error) {
        console.log(`Could not get posts created by the user: ${error}`);
      }
    }
    getUserPosts();
  }, [isLoggedIn, navigate]);

  useEffect(() => {
    async function fetchComments() {
      if (!isLoggedIn) {
        dispatch(toggleSignInModal());
        return;
      }
      try {
        const userCommentsRef = collection(db, "comments");

        const q = query(userCommentsRef, where("userId", "==", getUserId()));

        const userCommentsDoc = await getDocs(q);

        setCommentsPostId(userCommentsDoc.docs[0].data().postId);
        setUserComments(userCommentsDoc.docs);
      } catch (error) {
        console.log(`Could not fetch comments: ${error}`);
      }
    }
    fetchComments();
  }, [isLoggedIn, navigate]);

  useEffect(() => {
    async function fetchPostsFromUpVotes() {
      if (!isLoggedIn) {
        dispatch(toggleSignInModal());
        return;
      }
      try {
        const upVotesRef = collection(db, "users", `${getUserId()}/postVotes`);

        const q = query(upVotesRef, where("voteValue", "==", 1));

        const { docs } = await getDocs(q);

        const newDocs = docs.map(async (document: DocumentData) => {
          const postId = document.data().postId;
          const postRef = doc(db, "posts", postId);

          const data = await getDoc(postRef);

          return data;
        });

        const resolvedData = await Promise.all(newDocs);

        setUpVotesPosts(resolvedData);
      } catch (error) {
        console.log(`Could not fetch upvoted posts: ${error}`);
      }
    }

    fetchPostsFromUpVotes();
  }, [isLoggedIn]);

  useEffect(() => {
    async function fetchPostsFromDownVotes() {
      if (!isLoggedIn) {
        dispatch(toggleSignInModal());
        return;
      }
      try {
        const downVotesRef = collection(
          db,
          "users",
          `${getUserId()}/postVotes`
        );

        const q = query(downVotesRef, where("voteValue", "==", -1));

        const { docs } = await getDocs(q);

        const newDocs = docs.map(async (document: DocumentData) => {
          const postId = document.data().postId;

          const postRef = doc(db, "posts", postId);

          const data = await getDoc(postRef);

          return data;
        });

        const resolvedData = await Promise.all(newDocs);

        setDownVotesPosts(resolvedData);
      } catch (error) {
        console.log(`Could not fetch downvoted posts: ${error}`);
      }
    }

    fetchPostsFromDownVotes();
  }, [isLoggedIn]);

  useEffect(() => {
    async function fetchSavedPosts() {
      if (!isLoggedIn) {
        dispatch(toggleSignInModal());
        return;
      }

      try {
        const savedPostsRef = collection(db, "savedPosts");

        const q = query(savedPostsRef, where("userId", "==", getUserId()));

        const { docs } = await getDocs(q);

        const newDocs = docs.map(async (document: DocumentData) => {
          const postId = document.data().postId;

          const postRef = doc(db, "posts", postId);

          const data = await getDoc(postRef);

          return data;
        });

        const resolvedData = await Promise.all(newDocs);

        setSavedPosts(resolvedData);
      } catch (error) {
        console.log(`Could not fetch saved posts: ${error}`);
      }
    }

    fetchSavedPosts();
  }, [isLoggedIn]);

  const contentRef = useRef<HTMLDivElement>(null);

  return (
    <div styleName="profile-container">
      <Header dispatch={dispatch} activeSection={activeSection} />
      <Main>
        <div styleName="content" ref={contentRef}>
          {activeSection.posts && <Posts posts={userPosts} />}
          {activeSection.comments && (
            <Card>
              <Comments
                comments={userComments}
                commentsPostId={commentsPostId}
              />
            </Card>
          )}
          {activeSection.downvotes && <Posts posts={downVotesPosts} />}
          {activeSection.upvotes && <Posts posts={upVotesPosts} />}
          {activeSection.saved && <Posts posts={savedPosts} />}
          {activeSection.downvotes && downVotesPosts?.length == false && (
            <ProfileNotFound
              text={`hmm... looks like you haven't downvoted anything yet`}
            />
          )}
          {activeSection.upvotes && upVotesPosts?.length == false && (
            <ProfileNotFound
              text={`hmm... looks like you haven't upvoted anything yet`}
            />
          )}
          {activeSection.posts && userPosts?.length == false && (
            <ProfileNotFound
              text={`hmm... u/${getUserName()} hasn't posted anything`}
            />
          )}
          {activeSection.comments && userComments?.length == false && (
            <ProfileNotFound
              text={`hmm... u/${getUserName()} hasn't commented on anything`}
            />
          )}
          {activeSection.saved && savedPosts?.length == false && (
            <ProfileNotFound
              text={`hmm... u/${getUserName()} hasn't saved any posts`}
            />
          )}
        </div>
        <Aside>
          <UserCard />
        </Aside>
      </Main>
    </div>
  );
};

export default CSSModules(Profile, styles, {
  allowMultiple: true,
  handleNotFoundStyleName: "ignore",
});
