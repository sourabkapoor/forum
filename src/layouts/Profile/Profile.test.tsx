import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";
import Profile from "./Profile";

describe("Profile", () => {
  test("snapshot", () => {
    const { asFragment } = render(<Profile isLoggedIn={true} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
