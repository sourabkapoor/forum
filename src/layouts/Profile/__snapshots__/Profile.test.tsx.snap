// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile snapshot 1`] = `
<DocumentFragment>
  <div
    class="profile"
  >
    <button
      class="profile__button"
    >
      <span
        class="profile__details"
      >
        <img
          alt="default reddit avatar"
          class="profile__avatar"
          src="https://styles.redditmedia.com/t5_g2km0/styles/profileIcon_snood9e1c00c-b1c6-46a9-a231-a6fcd78cdd16-headshot.png?width=256&height=256&frame=1&crop=256:256,smart&s=7bd48be150588ab7dc3dc5c73be4d0dcddeeae8d"
        />
        <div
          class="profile__content"
        >
          <p
            class="profile__username"
          />
          <p
            class="profile__karma"
          >
            130 Karma
          </p>
        </div>
      </span>
      <svg
        class="profile__button-icon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 294.1L383 167c9.4-9.4 24.6-9.4 33.9 0s9.3 24.6 0 34L273 345c-9.1 9.1-23.7 9.3-33.1.7L95 201.1c-4.7-4.7-7-10.9-7-17s2.3-12.3 7-17c9.4-9.4 24.6-9.4 33.9 0l127.1 127z"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
